// ng-zorro-antd/bundles/ng-zorro-antd-spin.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('ng-zorro-antd/core'), require('@angular/cdk/observers'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-zorro-antd/spin', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', 'ng-zorro-antd/core', '@angular/cdk/observers', '@angular/common'], factory) :
    (global = global || self, factory((global['ng-zorro-antd'] = global['ng-zorro-antd'] || {}, global['ng-zorro-antd'].spin = {}), global.ng.core, global.rxjs, global.rxjs.operators, global['ng-zorro-antd'].core, global.ng.cdk.observers, global.ng.common));
}(this, function (exports, core, rxjs, operators, core$1, observers, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NzSpinComponent = /** @class */ (function () {
        function NzSpinComponent(cdr) {
            this.cdr = cdr;
            this.nzSize = 'default';
            this.nzDelay = 0;
            this.nzSimple = false;
            this.nzSpinning = true;
            this.loading = true;
            this.spinning$ = new rxjs.BehaviorSubject(this.nzSpinning);
            this.loading$ = this.spinning$.pipe(operators.debounceTime(this.nzDelay));
        }
        /**
         * @return {?}
         */
        NzSpinComponent.prototype.subscribeLoading = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.unsubscribeLoading();
            this.loading_ = this.loading$.subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this.loading = data;
                _this.cdr.markForCheck();
            }));
        };
        /**
         * @return {?}
         */
        NzSpinComponent.prototype.unsubscribeLoading = /**
         * @return {?}
         */
        function () {
            if (this.loading_) {
                this.loading_.unsubscribe();
                this.loading_ = null;
            }
        };
        /**
         * @return {?}
         */
        NzSpinComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            this.subscribeLoading();
        };
        /**
         * @param {?} changes
         * @return {?}
         */
        NzSpinComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
        function (changes) {
            if (changes.nzSpinning) {
                if (changes.nzSpinning.isFirstChange()) {
                    this.loading = this.nzSpinning;
                }
                this.spinning$.next(this.nzSpinning);
            }
            if (changes.nzDelay) {
                this.loading$ = this.spinning$.pipe(operators.debounceTime(this.nzDelay));
                this.subscribeLoading();
            }
        };
        /**
         * @return {?}
         */
        NzSpinComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            this.unsubscribeLoading();
        };
        NzSpinComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'nz-spin',
                        exportAs: 'nzSpin',
                        preserveWhitespaces: false,
                        encapsulation: core.ViewEncapsulation.None,
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        template: "<ng-template #defaultIndicatorTemplate>\n  <span class=\"ant-spin-dot\" [class.ant-spin-dot-spin]=\"loading\">\n    <i class=\"ant-spin-dot-item\"></i><i class=\"ant-spin-dot-item\"></i><i class=\"ant-spin-dot-item\"></i><i class=\"ant-spin-dot-item\"></i>\n  </span>\n</ng-template>\n<div *ngIf=\"loading\">\n  <div class=\"ant-spin\"\n    [class.ant-spin-spinning]=\"loading\"\n    [class.ant-spin-lg]=\"nzSize === 'large'\"\n    [class.ant-spin-sm]=\"nzSize === 'small'\"\n    [class.ant-spin-show-text]=\"nzTip\">\n    <ng-template [ngTemplateOutlet]=\"nzIndicator || defaultIndicatorTemplate\"></ng-template>\n    <div class=\"ant-spin-text\" *ngIf=\"nzTip\">{{ nzTip }}</div>\n  </div>\n</div>\n<div *ngIf=\"!nzSimple\"\n  class=\"ant-spin-container\"\n  [class.ant-spin-blur]=\"loading\">\n  <ng-content></ng-content>\n</div>\n",
                        host: {
                            '[class.ant-spin-nested-loading]': '!nzSimple'
                        },
                        styles: ["\n      nz-spin {\n        display: block;\n      }\n    "]
                    }] }
        ];
        /** @nocollapse */
        NzSpinComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef }
        ]; };
        NzSpinComponent.propDecorators = {
            nzIndicator: [{ type: core.Input }],
            nzSize: [{ type: core.Input }],
            nzTip: [{ type: core.Input }],
            nzDelay: [{ type: core.Input }],
            nzSimple: [{ type: core.Input }],
            nzSpinning: [{ type: core.Input }]
        };
        __decorate([
            core$1.InputNumber(),
            __metadata("design:type", Object)
        ], NzSpinComponent.prototype, "nzDelay", void 0);
        __decorate([
            core$1.InputBoolean(),
            __metadata("design:type", Object)
        ], NzSpinComponent.prototype, "nzSimple", void 0);
        __decorate([
            core$1.InputBoolean(),
            __metadata("design:type", Object)
        ], NzSpinComponent.prototype, "nzSpinning", void 0);
        return NzSpinComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NzSpinModule = /** @class */ (function () {
        function NzSpinModule() {
        }
        NzSpinModule.decorators = [
            { type: core.NgModule, args: [{
                        exports: [NzSpinComponent],
                        declarations: [NzSpinComponent],
                        imports: [common.CommonModule, observers.ObserversModule]
                    },] }
        ];
        return NzSpinModule;
    }());

    exports.NzSpinComponent = NzSpinComponent;
    exports.NzSpinModule = NzSpinModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
