// ng-zorro-antd/esm2015/message/nz-message-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { InjectionToken } from '@angular/core';
/**
 * @record
 */
export function NzMessageConfig() { }
if (false) {
    /** @type {?|undefined} */
    NzMessageConfig.prototype.nzAnimate;
    /** @type {?|undefined} */
    NzMessageConfig.prototype.nzDuration;
    /** @type {?|undefined} */
    NzMessageConfig.prototype.nzMaxStack;
    /** @type {?|undefined} */
    NzMessageConfig.prototype.nzPauseOnHover;
    /** @type {?|undefined} */
    NzMessageConfig.prototype.nzTop;
    /* Skipping unhandled member: [index: string]: any;*/
}
/** @type {?} */
export const NZ_MESSAGE_DEFAULT_CONFIG = new InjectionToken('NZ_MESSAGE_DEFAULT_CONFIG');
/** @type {?} */
export const NZ_MESSAGE_CONFIG = new InjectionToken('NZ_MESSAGE_CONFIG');
const ɵ0 = {
    nzAnimate: true,
    nzDuration: 3000,
    nzMaxStack: 7,
    nzPauseOnHover: true,
    nzTop: 24
};
/** @type {?} */
export const NZ_MESSAGE_DEFAULT_CONFIG_PROVIDER = {
    provide: NZ_MESSAGE_DEFAULT_CONFIG,
    useValue: ɵ0
};
export { ɵ0 };
