// ng-zorro-antd/esm2015/page-header/nz-page-header-cells.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive } from '@angular/core';
export class NzPageHeaderTitleDirective {
}
NzPageHeaderTitleDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-page-header-title, [nz-page-header-title]',
                exportAs: 'nzPageHeaderTitle',
                host: {
                    class: 'ant-page-header-title-view-title'
                }
            },] }
];
export class NzPageHeaderSubtitleDirective {
}
NzPageHeaderSubtitleDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-page-header-subtitle, [nz-page-header-subtitle]',
                exportAs: 'nzPageHeaderSubtitle',
                host: {
                    class: 'ant-page-header-title-view-sub-title'
                }
            },] }
];
export class NzPageHeaderContentDirective {
}
NzPageHeaderContentDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-page-header-content, [nz-page-header-content]',
                exportAs: 'nzPageHeaderContent',
                host: {
                    class: 'ant-page-header-content-view'
                }
            },] }
];
export class NzPageHeaderTagDirective {
}
NzPageHeaderTagDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-page-header-tags, [nz-page-header-tags]',
                exportAs: 'nzPageHeaderTags',
                host: {
                    class: 'ant-page-header-title-view-tags'
                }
            },] }
];
export class NzPageHeaderExtraDirective {
}
NzPageHeaderExtraDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-page-header-extra, [nz-page-header-extra]',
                exportAs: 'nzPageHeaderExtra',
                host: {
                    class: 'ant-page-header-title-view-extra'
                }
            },] }
];
export class NzPageHeaderFooterDirective {
}
NzPageHeaderFooterDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-page-header-footer, [nz-page-header-footer]',
                exportAs: 'nzPageHeaderFooter',
                host: {
                    class: 'ant-page-header-footer'
                }
            },] }
];
export class NzPageHeaderBreadcrumbDirective {
}
NzPageHeaderBreadcrumbDirective.decorators = [
    { type: Directive, args: [{
                selector: 'nz-breadcrumb[nz-page-header-breadcrumb]',
                exportAs: 'nzPageHeaderBreadcrumb'
            },] }
];
